### PR TITLE
lib/connections: Make assumptions about isLAN when interface address listing fails

### PR DESCRIPTION
**Pull request by bt90**
### Purpose

* always treat link-local IPs as LAN
* if we're unable to determine our local networks, simply treat private IPs as LAN

### Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

### CODE REVIEW

**Comment by calmh** [comment]
Link locals are quite unambiguously "LAN" so we could do that check early before running `GetLans` I think. The rest seems fine, maybe we should document somewhere what we mean with "LAN"...

**Comment by bt90** [comment]
I'd argue it would be the better solution because right now we treat all routed private IPs as WAN.

That might sound like an edge case but we might hit the samr problem with docker deployments.

**Comment by bt90** [comment]
My proposal would be to treat these as LAN:

* private IP ranges (RFC 1918 + RFC 4193)
* link-local IP ranges
* networks connected to our interfaces
* `AlwaysLocalNets`

There's no guarantee that a private IP is truly in your LAN and not routed via VPN across the globe. But that problem already exists today because we can't distinguish a physical interface from virtual one. So another node connected via Wireguard is also considered to be part of your LAN.

*If* we want to fix this, we'd need to probe latency.

**Comment by bt90** [comment]
> Only thing is, today one can configure Syncthing to match reality using a combination of detection + AlwaysLocalLans.

My other node is reachable via its Wireguard interface IP. The local node discovers `10.23.5.101/24` via its own network interface `wg0` and will treat my server with `10.10.0.102` as a LAN host.

**Comment by bt90** [comment]
At least for docker deployments the current logic doesn't work:

`ip addr show` inside the container yields:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
929: eth0@if930: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:ac:12:00:0f brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.15/16 brd 172.18.255.255 scope global eth0
       valid_lft forever preferred_lft forever
```

The user would need to explicitly specify e.g. `192.168.178.0/24` as part of `AlwaysLocalNets`.

**Comment by bt90** [comment]
The question is, what's worse in terms of dialing and connection upgrades?

If we misclassify a remote host connected via VPN as LAN, we only confuse the user, but still choose a good network path.

Misclassifying a LAN as a WAN IP can have negative consequences. The address ends up in the same priority group as the real WAN IP, which may be dialled first. Since both have the same priority, we're stuck with a sub-optimal network path.

**Comment by calmh** [comment]
So maybe a new option and `PrivateAddressSpaceCanBeWAN` which defaults to off. Then we can check IsLinkLocal first (always), IsPrivate (unless PrivateAddressSpaceCanBeWAN), and then AlwaysLocalNetworks and GetLans

**Comment by bt90** [comment]
`PrivateAddressesAlwaysLAN` with default on?

**Comment by bt90** [comment]
I've pulled the link-local check up. The PR should fix the problems we're currently facing on Android without having a negative impact on other platforms.

I've had a quick look at what's needed to introduce a new config knob and I'm not sure I'm up to the task :sweat_smile:

File: lib/connections/service.go [hunk 1]
```diff
@@ -793,7 +797,13 @@ func (s *lanChecker) isLAN(addr net.Addr) bool {
 		}
 	}
 
-	lans, _ := osutil.GetLans()
+	lans, err := osutil.GetLans()
+
+	if err != nil {
+		l.Debugln("Failed to retrieve interface IPs:", err)
+		return ip.IsPrivate()
+	}
```
Comment (calmh) [inline lib/connections/service.go:805]:
```suggestion
	lans, err := osutil.GetLans()
	if err != nil {
		l.Debugln("Failed to retrieve interface IPs:", err)
		priv := ip.IsPrivate()
		l.Debugln("Assuming", IP, "is local:", priv)
		return priv
	}
```

**Pull request merged**